app/src/main/java/nerd/tuxmobil/fahrplan/congress/details/SessionDetailsViewModel.kt

```diff
@@ -189,6 +195,18 @@ class SessionDetailsViewModel(
         setAlarm.postValue(Unit)
     }
 
+    fun tryToSetAlarm(context: Context, permissionLauncher: ActivityResultLauncher<String>, resourceString: String) {
```

Comment:
Please cover the logic of the view model in unit test(s). You will find examples for other public functions in `SessionDetailsViewModelTest`. Writting the test will probably show that:

1. The `NotificationHelper` should be injected via the `SessionDetailsViewModelFactory`.
2. The logic of this function can be hidden behind the existing `setAlarm()` function.
3. UI actions such as `Toast` should only happen in the corresponding `SessionDetailsFragment`. Therefore, you need to introduce observable properties similar to the existing ones in the view model.

Reply:
> Please cover the logic of the view model in unit test(s).

I would do that, the Problem is that there isn't really anything I can cover here (as far as I know). The Logic is mostly just calling Methods of other Classes that should already be covered by unit tests, none of the Methods has a return value that can be checked, and the split between the `Fragment` and `ViewModel` classes makes it harder to test everything (again, as far as I know, which does not mean much as I don't have much experience with Unittests)

> The NotificationHelper should be injected via the SessionDetailsViewModelFactory

I tried that, but Android Studio kept yelling at me for either leaking a context objekt or exposing the `NotificationHelper` type

> The logic of this function can be hidden behind the existing setAlarm() function.

Will be changed in next push :+1: 

> UI actions such as Toast should only happen in the corresponding SessionDetailsFragment. Therefore, you need to introduce observable properties similar to the existing ones in the view model.

Will be changed in next push :+1:

Reply:
Here is how you can inject the `NotificationHelper` (untested):

``` kotlin
// SessionDetailsFragment

private lateinit var appRepository: AppRepository
private lateinit var alarmServices: AlarmServices
private lateinit var notificationHelper: NotificationHelper
private val viewModel: SessionDetailsViewModel by viewModels {
    SessionDetailsViewModelFactory(
        appRepository = appRepository,
        alarmServices = alarmServices,
        notificationHelper = notificationHelper,
        defaultEngelsystemRoomName = AppRepository.ENGELSYSTEM_ROOM_NAME,
        customEngelsystemRoomName = getString(R.string.engelsystem_shifts_alias)
    )
}
```
...

``` kotlin
// SessionDetailsFragment

override fun onAttach(@NonNull context: Context) {
    super.onAttach(context)
    appRepository = AppRepository
    alarmServices = AlarmServices.newInstance(context, appRepository)
    notificationHelper = NotificationHelper(context)
```

``` kotlin
// SessionDetailsViewModelFactory

internal class SessionDetailsViewModelFactory(

    private val appRepository: AppRepository,
    private val alarmServices: AlarmServices,
    private val notificationHelper: NotificationHelper,
    private val defaultEngelsystemRoomName: String,
    private val customEngelsystemRoomName: String

) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        @Suppress("UNCHECKED_CAST")
        return SessionDetailsViewModel(
            repository = appRepository,
            executionContext = AppExecutionContext,
            alarmServices = alarmServices,
            notificationHelper = notificationHelper,
            sessionFormatter = SessionFormatter(),
```


``` kotlin
// SessionDetailsViewModel

internal class SessionDetailsViewModel(

    private val repository: AppRepository,
    private val executionContext: ExecutionContext,
    private val alarmServices: AlarmServices,
    private val notificationHelper: NotificationHelper,
    private val sessionFormatter: SessionFormatter,
    private val simpleSessionFormat: SimpleSessionFormat,
```

I also see the Lint warning that `NotificationHelper` leaks the context but I think that is a false-positive here.

Reply:
The SDK check is logic which might better be part of the view model.

Reply:
Implemented :+1:

Reply:
> again, as far as I know, which does not mean much as I don't have much experience with Unittests

Let me know if you need help.

Reply:
Now explained in https://github.com/isiko/EventFahrplan/pull/1.